### PR TITLE
Support MXFP8 batch invariance for inference and TE grouped training

### PR DESCRIPTION
**Pull request by wdykas**
Add MXFP8 batch-invariant inference for Torch/vLLM and FlashInfer, plus TE device-metadata grouped GEMM training. This PR targets `main`; the selective-precision prerequisite #7300 has already merged.

## Summary

- Support MXFP8 batch-invariant inference for Torch and vLLM-selected MXFP8 layers using MCore quantization and PyTorch scaled grouped GEMM. Both squared-ReLU and SwiGLU experts are supported.
- Support FlashInfer routed MXFP8 batch-invariant inference for non-gated squared-ReLU experts, including selective BF16 expert layers.
- Align and zero-fill expert padding before FC1/FC2 quantization; disable fused quantization when the batch-invariant inverse map is needed. Match training-policy SwiGLU probability/rounding order and use 64-bit activation/permutation addressing.
- Use shared RoPE in batch-invariant mode; reject an explicit FlashInfer fused-RoPE request.
- Support TE device-metadata grouped GEMM for training via `moe_use_grouped_tensor=True`, with 256-row expert padding. Its full workspace is separate for TN/NN/NT layouts and gated to SM100 + cuBLASLt 13.5.1 with TE's `GroupedLinear(use_grouped_tensor=...)` API. Dense and legacy grouped GEMMs retain restricted workspace. TE op fuser remains unsupported, and the legacy environment switch alone is insufficient.
- Preserve #7300's construction-time precision selection and storage-based refit. Native TE inference is outside this PR.

## Validation

After merging current `main` in `860a43c11`, the diff contains only the original 17-file change. Sixteen files are identical to the previously reviewed versions; `transformer_config.py` also preserves the upstream MTP validation changes.

Validation on `860a43c11`: six focused configuration tests passed; the determinism coverage gate against `main`, Black, isort, Ruff, and diff checks passed. The GPU suite below was not repeated for this merge.

Original implementation validation on `691d03eaa`:

GB200, PyTorch 2.11.0+cu130, TE `2.20.0.dev0+51317b4`, cuBLASLt 13.5.1:

- 984 unit/configuration/FlashInfer/workspace/selective-precision tests passed; 7 optional-dependency tests skipped.
- 24 selected determinism tests passed, including ten actual TE grouped forward/backward/CUDA-graph cases (BF16/MXFP8, changing expert sizes, empty experts, and Nano dimensions). One existing non-strict racy-unpermute negative control XPASS; 17 deselected.
- Six two-GPU EP/NVLS training-vs-inference MoE parity cases passed at `atol=0, rtol=0`, including TE device-metadata training versus Torch/vLLM-selected MXFP8 inference.
- Four GPT dynamic/batched-forward and co-batch invariance tests passed.
- Black, isort, Ruff, and diff check passed. The determinism coverage gate was rechecked against #7300 for the consolidated 17-file diff and passed.

The previous Qwen/Nano NeMo RL zero-logprob smoke results are documented in #7333. Those full-model RL runs were not repeated for this combination; kernel/layer regression results are not a general full-model parity guarantee.

## Numerical scope

FlashInfer routed MXFP8 is bitwise invariant across batch composition, but its GEMM is not bitwise identical to the TE training-policy GEMM. The earlier direct two-rank comparison differed in approximately 96% of one-layer BF16 outputs with maximum absolute error 8.5449e-4. Use Torch/vLLM-selected MXFP8 inference for the validated exact training-policy comparisons.

Forward invariance does not promise identical gradients across different minibatches.

### CODE REVIEW

File: megatron/core/inference/moe/activations.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def _squared_relu_kernel(
     CLAMP: tl.constexpr,
     BLOCK_N: tl.constexpr,
     NUM_BLOCKS: tl.constexpr,  # grid size (fixed for CG)
+    ZERO_PADDING: tl.constexpr,
```
Comment (wdykas) [inline megatron/core/inference/moe/activations.py:71]: make notes of why we need 0 padding in this for quantization.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 2]
```diff
@@ -860,8 +860,12 @@ def __init__(self, model_config: TransformerConfig, inference_config: InferenceC
                 f"ssm_chunk_alignment ({self.ssm_chunk_alignment})."
             )
 
-        # FlashInfer.
-        if inference_config.use_flashinfer_fused_rope is True:
+        # FlashInfer's fused RoPE does not produce the same bits as the RoPE path
+        # used by a policy forward. Batch-invariant mode promises parity between
+        # dynamic generation and that forward, so keep both on the shared path.
```
Comment (wdykas) [inline megatron/core/inference/contexts/dynamic_context.py:865]: decide if the silent behavior is fine or we should just assert false

File: megatron/core/inference/moe/activations.py (hunk 1 above)
Comment (wdykas) [reply megatron/core/inference/moe/activations.py:71]: Documented this explicitly: MXFP8 quantization and grouped GEMM do not receive the permutation map, so aligned dummy rows must be materialized as neutral zeros before those operations.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 2 above)
Comment (wdykas) [reply megatron/core/inference/contexts/dynamic_context.py:865]: Tightened this contract: an explicit use_flashinfer_fused_rope=True now raises a clear ValueError in batch-invariant mode; None still resolves to False so automatic selection safely uses the shared policy-compatible RoPE path.

File: megatron/core/transformer/custom_layers/batch_invariant_kernels.py [hunk 3]
```diff
@@ -1718,6 +1737,18 @@ def _enable_te_native_workspace_starvation(workspace_bytes: int = _TE_NATIVE_WOR
             _te_gemm_mod.get_cublas_workspace_size_bytes = lambda: workspace_bytes
             if hasattr(getattr(_te_gemm_mod, "get_cublas_workspace", None), "cache_clear"):
                 _te_gemm_mod.get_cublas_workspace.cache_clear()
+        # TE's device-metadata grouped GEMM rejects a workspace smaller than its normal
+        # allocation. This path gets its batch invariance from padding every expert segment
+        # to a 256-row multiple, so preserve the mandatory full workspace at this narrow
+        # entry point while keeping ordinary GEMMs starved.
+        if _TE_GROUPED_WORKSPACE_FN_ORIG is None and hasattr(
```
Comment (wdykas) [inline megatron/core/transformer/custom_layers/batch_invariant_kernels.py:1737]: I think we gate on mxfp8 here

File: tests/unit_tests/determinism/kernels/manifest.py [hunk 4]
```diff
@@ -452,11 +452,12 @@ class KernelEntry:
         sources=(
             "megatron/core/inference/moe/activations.py",
             "megatron/core/inference/moe/batch_invariant.py",
+            "megatron/core/inference/moe/fused_moe.py",
             "megatron/core/inference/moe/permute.py",
             "megatron/core/inference/moe/vllm_fused_moe.py",
             "megatron/core/inference/quantization/mxfp8_quantize.py",
         ),
-        tests=(K + "test_inference_kernels.py",),
+        tests=(K + "test_inference_kernels.py", "tests/unit_tests/inference/test_mxfp8_utils.py"),
```
Comment (wdykas) [inline tests/unit_tests/determinism/kernels/manifest.py]: todo: fix

File: megatron/core/transformer/moe/experts.py [hunk 5]
```diff
@@ -318,9 +330,12 @@ def __init__(
...
             assert HAVE_TE, "Quantized or TE grouped-tensor GroupedMLP execution requires TE."
-            align_size = (
-                get_align_size_for_quantization(self.config) if self._use_grouped_tensor else None
-            )
+            # Older TE releases select their device-metadata grouped-tensor path only through
+            # this environment variable (there is no constructor argument for MCore to mirror
+            # into ``moe_use_grouped_tensor``). That path requires 256-row expert segments,
+            # while ordinary MXFP8 padding is only 32 rows. Over-pad when either interface
+            # requests grouped tensors so the metadata and physical rows remain consistent.
+            align_size = _te_grouped_tensor_align_size(self.config)
```
Comment (wdykas) [inline megatron/core/transformer/moe/experts.py]: why are we padding to 256 as the minimum here.  This should be gated on batch invariance

File: megatron/core/transformer/custom_layers/batch_invariant_kernels.py (hunk 3 above)
Comment (wdykas) [reply megatron/core/transformer/custom_layers/batch_invariant_kernels.py:1737]: I removed the workspace exemption instead. The guard covers BF16 too, since mixed-precision recipes can hit the same grouped-tensor path.

File: tests/unit_tests/determinism/kernels/manifest.py (hunk 4 above)
Comment (wdykas) [reply tests/unit_tests/determinism/kernels/manifest.py]: Moved the co-batch test into the determinism kernel suite and removed the duplicate manifest entries. The coverage check passes.

File: megatron/core/transformer/moe/experts.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/transformer/moe/experts.py]: Removed the extra padding helper. This PR no longer changes training alignment; the unsupported grouped-tensor path now raises in batch-invariant mode.

File: tests/unit_tests/inference/test_moe_dispatching_and_routing.py [hunk 6]
```diff
@@ -733,6 +775,14 @@ def test_batch_invariant_moe_matches_training(
...
             # parametrization keeps its existing coverage.
             use_fused_weighted_squared_relu=activation_clamp_scale is not None,
+            gated_linear_unit=mxfp8_swiglu,
+            activation_func=(torch.nn.functional.silu if mxfp8_swiglu else squared_relu),
+            bias_activation_fusion=mxfp8_swiglu,
+            fp8=("hybrid" if mxfp8_swiglu else None),
+            fp8_recipe=("mxfp8" if mxfp8_swiglu else "delayed"),
+            fp8_param=mxfp8_swiglu,
+            use_cpu_initialization=not mxfp8_swiglu,
+            batch_invariant_backend=("te_native" if mxfp8_swiglu else None),
```
[the remark this answers is not in the document]
Comment (wdykas) [reply tests/unit_tests/inference/test_moe_dispatching_and_routing.py]: Fixed — the config now uses `te_native`. The separate runtime backend selection is unchanged.

File: megatron/core/inference/moe/permute.py [hunk 7]
```diff
@@ -295,6 +295,30 @@ def _permute_tokens_kernel(
...
+    pid = tl.program_id(0)
+    n_used = tl.load(n_used_ptr)
+    if pid >= n_used:
+        return
+    for row in tl.range(pid, max_rows, NUM_BLOCKS):
+        if row < n_used:
+            if tl.load(permutation_map_ptr + row) < 0:
+                for h in tl.range(0, hidden_dim, BLOCK_H):
+                    offsets = h + tl.arange(0, BLOCK_H)
+                    mask = offsets < hidden_dim
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/inference/moe/permute.py:318]: Made the hidden-buffer offsets int64 in the neighboring permute/unpermute kernels too, including fused MXFP8 quantization.

File: megatron/core/transformer/transformer_config.py [hunk 8]
```diff
@@ -3553,24 +3553,50 @@ def _scope_to_str(s):
...
                 )
-                assert not (
-                    self.fp8 or self.fp4
-                ), "Batch-invariant MoE is bf16-only. Disable fp8/fp4 to use it."
+                squared_relu_or_swiglu = (
+                    not self.gated_linear_unit and self.activation_func == squared_relu
+                ) or (self.gated_linear_unit and self.activation_func == F.silu)
+                if self.transformer_impl == "inference_optimized":
+                    mxfp8_supported = (
+                        mxfp8_params_enabled
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/transformer_config.py:3552]: Updated both sections with the MXFP8 support and the legacy TE training-path restriction.

File: megatron/core/transformer/custom_layers/batch_invariant_kernels.py [hunk 9]
```diff
@@ -1718,6 +1737,18 @@ def _enable_te_native_workspace_starvation(workspace_bytes: int = _TE_NATIVE_WOR
...
+            _te_gemm_mod, "_get_grouped_cublas_workspace"
+        ):
+            _TE_GROUPED_WORKSPACE_FN_ORIG = _te_gemm_mod._get_grouped_cublas_workspace
+            if hasattr(_TE_GROUPED_WORKSPACE_FN_ORIG, "cache_clear"):
+                _TE_GROUPED_WORKSPACE_FN_ORIG.cache_clear()
+            _get_unrestricted_te_grouped_workspace.cache_clear()
+            _te_gemm_mod._get_grouped_cublas_workspace = _get_unrestricted_te_grouped_workspace
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/custom_layers/batch_invariant_kernels.py]: You're right: alignment doesn't fix M. I took option 1 — removed the full-workspace exemption and reject device-metadata grouped GEMM in batch-invariant mode. Legacy grouped GEMM keeps the restricted workspace; the exact parity results are layer-level, not full-model logprob parity.

File: megatron/core/transformer/transformer_config.py [hunk 10]
```diff
@@ -3553,24 +3553,50 @@ def _scope_to_str(s):
...
+                        in (InferenceGroupedGemmBackend.TORCH, InferenceGroupedGemmBackend.VLLM)
+                        and squared_relu_or_swiglu
+                    )
+                else:
+                    # The training policy uses TE GroupedLinear directly; the inference
+                    # backend selector is generation-only and therefore irrelevant here.
+                    mxfp8_supported = (
+                        mxfp8_params_enabled
+                        and self.moe_grouped_gemm
+                        and self.batch_invariant_backend == "te_native"
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/transformer_config.py:3563]: The exemption and alignment helper are gone. The config now rejects grouped-tensor/fused-grouped training under `te_native`, with a dispatch guard as well; legacy MXFP8 grouped GEMM keeps its normal alignment and restricted workspace.

File: megatron/core/transformer/moe/experts.py [hunk 11]
```diff
@@ -94,6 +95,17 @@
...
 
+def _te_grouped_tensor_align_size(config: TransformerConfig) -> Optional[int]:
+    """Return explicit padding for TE's device-metadata grouped-linear path."""
+    requested = (
+        config.moe_use_grouped_tensor
+        or os.getenv("NVTE_GROUPED_LINEAR_USE_FUSED_GROUPED_GEMM", "0") == "1"
+    )
+    if not requested:
+        return None
+    return max(get_align_size_for_quantization(config), 256)
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/moe/experts.py]: Removed the env-driven padding change entirely. `experts.py` is now unchanged relative to the base PR.

File: megatron/core/inference/moe/activations.py [hunk 12]
```diff
@@ -87,13 +90,18 @@ def _squared_relu_kernel(
                     x = tl.load(input_ptr + row * N + o, mask=m).to(tl.float32)
                     r = _clamped_relu(x, clamp_scale, CLAMP)
                     tl.store(output_ptr + row * N + o, (r * r).to(tl.bfloat16), mask=m)
+            elif ZERO_PADDING:
+                for n in tl.range(0, N, BLOCK_N):
+                    o = n + tl.arange(0, BLOCK_N)
+                    tl.store(output_ptr + row * N + o, 0.0, mask=o < N)
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/inference/moe/activations.py]: Fixed the sibling kernels too: hidden-buffer loads/stores and MXFP8 scale offsets now promote before the large multiplications. The activation and permutation tests pass.

File: megatron/core/inference/moe/fused_moe.py [hunk 13]
```diff
@@ -161,18 +161,29 @@ def mcore_fused_moe(
...
+            assert (
+                HAVE_SCALED_GMM
+            ), "Torch MXFP8 inference requires torch.nn.functional.scaled_grouped_mm."
+            mm_fn = _mxfp8_grouped_mm
+            # Keep every expert boundary aligned to the MXFP8 scale swizzle.
+            expert_alignment = MXFP8_SCALE_ROW_BLOCK
+        else:
+            mm_fn = batch_invariant.grouped_mm
+            expert_alignment = batch_invariant.grouped_mm_alignment()
     elif use_mxfp8:
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/inference/moe/fused_moe.py]: Done — one MXFP8 branch now handles both modes, using `MXFP8_SCALE_ROW_BLOCK` and a single availability check.

File: megatron/core/transformer/custom_layers/batch_invariant_kernels.py [hunk 14]
```diff
@@ -1659,9 +1660,27 @@ def is_batch_invariant_mode_enabled():
...
+    return workspace_size_fn()
+
+
+@lru_cache(maxsize=None)
+def _get_unrestricted_te_grouped_workspace(device: int, layout: str) -> torch.Tensor:
+    """Allocate the full workspace required by TE's device-metadata grouped GEMM."""
+    assert layout in ("TN", "NN", "NT"), f"unexpected grouped GEMM layout {layout}"
+    return torch.empty(
+        _get_unrestricted_te_workspace_size_bytes(), dtype=torch.uint8, device=device
+    )
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/custom_layers/batch_invariant_kernels.py]: Removed this workspace allocator entirely, so the layout assertion and shared cache are gone. Unsupported grouped-tensor calls now raise a clear error.

**Review by yaox12 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by mathemakitten — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [860a43c](https://github.com/NVIDIA/Megatron-LM/pull/7302/commits/860a43c115789bf796a9780109e02d59c87ad0a3)
